### DISCUSSION TITLE
Simplify string interpolation (to help lexing)

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#6530] [Ideas]
Right now, you can have an arbitrary expression inside of a string interpolation, like so:
```flix
def example(): String = "hello ${"cool" + " p${"e"}ople"}, what's up?"
```
This makes lexing flix considerably more complex. Is this worth it? We could:
- Make interpolation support variables, `${x}`, `${y}`
- Make interpolation support simple literals `${12}`, `${true}`
- Make interpolation support expressions without `}`
- Avoid the issue by picking a more complicated end signal, like `${}$` and disallowing nesting, but then we probably need to support `$x`
- Some combination thereof

**Comment by magnus-madsen** [comment]
@herluf-ba Thoughts?

What if we disallow string interpolation inside string interpolation?

**Reply by JonathanStarup** [reply]
Even if you don't allow nested interpolation, you still have to count `}`'s to know where the interpolation ends. (`"${List#{1,2}}"`)

I believe the least restrictive rule is that interpolation cannot contain `}` (except in comments and strings). This is, of course a weird rule, but that's the minimum requirement.

**Reply by magnus-madsen** [reply]
Guh! Nice example.

**Reply by JonathanStarup** [reply]
I've added a possible solution, to have `${}$` so the end is unique to interpolation. Then disallowing nesting would be sufficient

**Reply by JonathanStarup** [reply]
Also, my goal was constant lookahead lexing, and just disallowing nested interpolation doesn't help that, but it does probably still make the lexer simpler

**Comment by mlutze** [comment]
For what it's worth, it looks like Rust effectively only supports variables.

**Reply by JonathanStarup** [reply]
True. I think it can be argued that the complexity of implementation is greater than the payoff.

**Comment by herluf-ba** [comment]
Allowing any expression to appear in an interpolation means that we need to call the top-level `scanToken` function again from within the string interpolation. Allowing nesting means that this can lead to indirect recursion, so then we need to keep track of the nesting level to avoid blowing the stack. Having a unique terminator for interpolations like `}$` would eliminate the need to count `}`s specifically, but we would still need to count the nesting level to make sure the stack is good.

To me, the main contributor of complexity is allowing any expression to appear in a string interpolation. This introduces the need to call `scanToken` again, but from a different call site, meaning that there is now some extra context. 
If we were to only allow names and literals for instance, all this complexity goes away. We could even simplify syntax a bit, so no curly's are needed. So then if a string needs to be computed, you just declare a variable with it before the interpolation.

```
let name = getName();
let emailStatus = if (numEmailsUnread > 0) {
    "You have $numEmailsUnread unread emails."
    } else {
        "You are all caught up"
    };
"Hi there $name! $emailStatus"
```

**Reply by herluf-ba** [reply]
I don't think we even need to support literals. Any literal can just be written directly in the string `"There are $5 kings in the north" -> "There are 5 kings in the north"`. So only names are needed really.

Another benefit of being restrictive and only allowing names is that it is very easy to communicate how the feature works. Something like:
> Any string can be interpolated by prefixing a variable with `$` within the string. No expressions are allowing within strings, instead declare a new variable and use that.

Now the behavior of the feature is easy to describe and understand.

**Reply by magnus-madsen** [reply]
Five kings in the North? That seems pretty unsustainable!

**Reply by herluf-ba** [reply]
A war is brewing for sure.

**Reply by mlutze** [reply]
If I'm reading that code I assume it means they're selling kings for 5 dollars :)

**Comment by JonathanStarup** [comment]
My vote (if it matters): Only allow variables (`${x}`). We can add the shorthand `$x` later. We can do something smarter later if we find some concrete use cases.

Another way to fail is if we have some complication rule and non-nesting. If we want anything better than a naive lexer/parser error then we need to lex it anyway in order to say its not allowed

**Comment by stephentetley** [comment]
Can I make a feature request for multiline strings with some interpolation at this point? (can be as simple as just interpolating variables). 

This would be very nice inlining SQL queries.

**Reply by magnus-madsen** [reply]
Yes, we should do this.
